WooCommerce/Classes/ViewRelated/Products/ProductsViewControllerStateCoordinator.swift

```diff
@@ -0,0 +1,68 @@
+import UIKit
+
+/// UI state of `ProductsViewController`.
+///
+/// - noResultsPlaceholder: a no results placeholder is displayed
+/// - syncing: syncing data UI with a parameter that indicates whether there is existing data
+/// - results: the results are shown
+enum ProductsViewControllerState {
```

Comment:
This seems like something that could be generalised, as it seems like those three states (syncing, no results and results) can apply to basically any view controller. Do you think this could be made more abstract? (maybe it's just a matter of renaming it to LoadingState / LoadingStateCoordinator)

Reply:
I looked at all the view controllers with view controller state before this PR (`OrdersViewController`, `OrderLoaderViewController`, `OrderSearchViewController`, `NotificationsViewController`, `ReviewsViewController`, `ProductLoaderViewController`) and all of them have different state cases - like some would have success state with a model, while some have syncing state with different parameters. I think having the view controller state per view controller would be more flexible. If we want to generalize it, maybe we could chat with the design and refactor for all the existing view controllers?

Reply:
🤔 thank we should probably extract this to a proper state machine. The thing is, if we could find a way to remove state management from view controllers, we would also remove a lot of code from all those view controllers as well. Let's give it some thought and try to find a solution in the nearly future. 

Thanks for looking into it!

Reply:
Created an issue to track this: https://github.com/woocommerce/woocommerce-ios/issues/1305